### PR TITLE
fix(include): fix rare bug involving subqueries on combination of multiple includes

### PR DESCRIPTION
**Pull request by wertlex**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones) (no updates)?
- [x] Did you update the typescript typings accordingly (if applicable)? (not applicable)
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change
Fixes https://github.com/sequelize/sequelize/issues/9035
Fixes abstract query generator, so all supported database engines are affected.

### CODE REVIEW

**Comment by papb** [comment]
Hello! I see you are a first-time contributor, thank you for taking the time to help Sequelize! I hope to see more PRs from you in the future!

File: test/integration/include/findAll.test.js [hunk 1]
```diff
@@ -2091,5 +2091,23 @@ describe(Support.getTestDialectTeaser('Include'), () => {
         });
       });
     });
+
+    it('should be able to generate a correct request for entity with 1:n and m:1 associations and limit', function() {
+      return this.fixtureA().then(() => {
+        return this.models.Product.findAll({
+          where: {},
```
Comment (papb) [inline test/integration/include/findAll.test.js]: Is this `where: {}` necessary?

File: test/integration/include/findAll.test.js [hunk 2]
```diff
@@ -2091,5 +2091,23 @@ describe(Support.getTestDialectTeaser('Include'), () => {
...
+          attributes: ['title'],
+          include: [
+            { model: this.models.User },
+            { model: this.models.Price }
+          ],
+          limit: 10
+        }).then( products => {
+          // checking that internally added fields used to handle 'BelongsTo' associations are not leaked to result
+          expect(products[0].UserId).to.be.equal(undefined);
+          expect(products[0].title).to.be.a('string');
```
Comment (papb) [inline test/integration/include/findAll.test.js]: Can you add more `expect`s to guarantee the expected query result? Just to make the test a bit more robust. Checking only the type of `title` of the first result seems weak

File: lib/dialects/abstract/query-generator.js [hunk 3]
```diff
@@ -1170,6 +1170,16 @@ class QueryGenerator {
           attributes.main.push(mainTable.model.rawAttributes[keyAtt].field ? [keyAtt, mainTable.model.rawAttributes[keyAtt].field] : keyAtt);
         }
       }
+
+      // appending fields which are required by 'BelongsTo' associations when main attributes filter is in use
+      if (options.include) {
+        for (const include of options.include) {
+          const { associationType, foreignKey } = include.association;
```
Comment (papb) [inline lib/dialects/abstract/query-generator.js]: I am not very familiar with this part of the code, how can we be 100% sure that `include.association` will always be defined here (and be an object)? Recall that `const { x, y } = undefined;` gives `TypeError: Cannot destructure property `x` of 'undefined' or 'null'`. I would be more confident if this was wrapped in an `if (include.association)`... What do you think?

File: test/integration/include/findAll.test.js (hunk 1 above)
Comment (wertlex) [reply test/integration/include/findAll.test.js]: Yeah, there was no need in it.

File: test/integration/include/findAll.test.js (hunk 2 above)
Comment (wertlex) [reply test/integration/include/findAll.test.js]: I've added a bit more `expect`s to check `products` shape more precisely.

File: lib/dialects/abstract/query-generator.js (hunk 3 above)
Comment (wertlex) [reply lib/dialects/abstract/query-generator.js]: Good point. Updated it to _paranoid_ mode

Comment (papb) [reply lib/dialects/abstract/query-generator.js]: @wertlex Excellent!

**Comment by papb** [comment]
Nice! Another question: you claim to be fixing #9035 (and I believe you), but it is not immediately clear to me from the PR modifications themselves. The OP from #9035 posted something about `defaultScope` being relevant to the bug, and I don't see anything on your PR about this. Can you explain to me a bit further what exactly is going on?

**Comment by wertlex** [comment]
Sure!
It looks like #9035 's topic starter problem is not in `scopes` itself, but in `attributes` value set in scopes. Note **Got this Error** part: "SequelizeDatabaseError: Unknown column 'pm_users.enterpriseId' in 'on clause'". Error is related not to scope, but to missed column.
So, skipping scopes and getting down to queries itself.

Let's consider Sequelize and corresponding SQL queries from test: https://github.com/sequelize/sequelize/pull/11385/files#diff-916a03a19d49cecbcd62820c088b9587R2095

Will start from something a bit simpler. Given Sequelize query: 
```
 return this.models.Product.findAll({
   attributes: ['title'],
   include: [
     { model: this.models.User },
     { model: this.models.Price }
   ]
})
```

will produce following SQL:
```
SELECT 
  "Product"."id", "Product"."title", "User"."id" AS "User.id", "User"."createdAt" AS "User.createdAt", "User"."updatedAt" AS "User.updatedAt", "Prices"."id" AS "Prices.id", "Prices"."value" AS "Prices.value", "Prices"."createdAt" AS "Prices.createdAt", "Prices"."updatedAt" AS "Prices.updatedAt", "Prices"."ProductId" AS "Prices.ProductId" 
  FROM "Products" AS "Product" 
    LEFT OUTER JOIN "Users" AS "User" ON "Product"."UserId" = "User"."id" 
    LEFT OUTER JOIN "Prices" AS "Prices" ON "Product"."id" = "Prices"."ProductId";
```

Important thing here is that it does not use subquery - it is just quite straight `SELECT * FROM` query with couple of joins.

But obviously it cannot handle `limit` well. So when Sequelize query stuffed with `limit`:
```
 return this.models.Product.findAll({
   attributes: ['title'],
   include: [
     { model: this.models.User },
     { model: this.models.Price }
   ],
   limit: 10
})
```
SQL query turned into a little bit different thing:
```
SELECT 
  "Product".*, "User"."id" AS "User.id", "User"."createdAt" AS "User.createdAt", "User"."updatedAt" AS "User.updatedAt", "Prices"."id" AS "Prices.id", "Prices"."value" AS "Prices.value", "Prices"."createdAt" AS "Prices.createdAt", "Prices"."updatedAt" AS "Prices.updatedAt", "Prices"."ProductId" AS "Prices.ProductId" 
  FROM (SELECT "Product"."id", "Product"."title" FROM "Products" AS "Product" LIMIT 10) AS "Product" 
    LEFT OUTER JOIN "Users" AS "User" ON "Product"."UserId" = "User"."id" 
    LEFT OUTER JOIN "Prices" AS "Prices" ON "Product"."id" = "Prices"."ProductId";
```
Note that subquery at `FROM (SELECT ...)` - it have `LIMIT 10` set as expected, but also it enumerates attributes passed in `attributes` property of Sequelize query. But next line (`LEFT OUTER JOIN "Users" AS "User" ON "Product"."UserId" = "User"."id"`) requires `"Product"."UserId"` field to perform join. And this field obviously is not in fields obtained for `Products`, which is effectively gives broken query.

This pr detects foreign keys for such relationships and pushes them into collection of attributes of main query, similar to primary key (which is also not passed in `attributes`, still appears in SQL query but not in result)

**Comment by sushantdhiman** [comment]
Thanks for this PR, I have investigated this issue and applied a different fix https://github.com/sequelize/sequelize/pull/11404

Please let me know if that works for you!

**Comment by wertlex** [comment]
Yeah, thank you. 
It works now :)

P.S. sorry for late reply

**Pull request closed without merging**